cpu/atmega_common/periph/adc.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2016 Laurent Navet <laurent.navet@gmail.com>
+ *               2017 HAW Hamburg, Dimitri Nahm
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     drivers_periph
+ * @{
+ *
+ * @file
+ * @brief       Low-level ADC driver implementation for ATmega family
+ *
+ * @author      Laurent Navet <laurent.navet@gmail.com>
+ * @author      Dimitri Nahm <dimitri.nahm@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "mutex.h"
+#include "assert.h"
+#include "periph/adc.h"
+#include "periph_conf.h"
+
+#define ADC_MAX_CLK         (200000U)
+
+static mutex_t lock = MUTEX_INIT;
+
+static inline void prep(void)
+{
+    mutex_lock(&lock);
+    /* Enable ADC */
+    ADCSRA |= (1 << ADEN);
+}
+
+static inline void done(void)
+{
+    /* Disable ADC */
+    ADCSRA &= ~(1 << ADEN);
+    mutex_unlock(&lock);
+}
+
+int adc_init(adc_t line)
+{
+    uint32_t clk_div = 1;
+
+    /* check if the line is valid */
+    if (line >= ADC_NUMOF) {
+        return -1;
+    }
+
+    prep();
+
+    /* Disable corresponding Digital input */
+    #if defined (CPU_ATMEGA328P) || defined (CPU_ATMEGA1281)
+        DIDR0 |= (1 << line);
+    #elif defined (CPU_ATMEGA2560)
+        if (line < 8)
+            DIDR0 |= (1 << line);
+        else
+            DIDR2 |= (1 << (line - 8));
+    #endif
+
+    /* Set ADC-pin as input */
+    #if defined (CPU_ATMEGA328P)
+        DDRC &= ~(1 << line);
+        PORTC &= ~(1 << line);
+    #elif defined (CPU_ATMEGA2560)
+        if (line < 8) {
+            DDRF &= ~(1 << line);
+            PORTF &= ~(1 << line);
+        }
+        else {
+            DDRK &= ~(1 << (line-8));
+            PORTK &= ~(1 << (line-8));
+        }
+    #elif defined (CPU_ATMEGA1281)
+        DDRF &= ~(1 << line);
+        PORTF &= ~(1 << line);
+    #endif
+
+    /* set clock prescaler to get the maximal possible ADC clock value */
+    for (clk_div = 1; clk_div < 8; clk_div++) {
+        if ((CLOCK_CORECLOCK / (1 << clk_div)) <= ADC_MAX_CLK) {
+            break;
+        }
+    }
+    ADCSRA |= clk_div;
+
+    /* Ref Voltage is Vcc(5V) */
+    ADMUX |= (1 << REFS0);
```

Comment:
I wouldn't force it to 5V, it' might be interesting to have VREF as reference voltage. Plus in the 1281/2560 it can be set to differential measurements which might be quite useful. 

I suggest introduce a configuration struct, which allows to configure it for each channel. Then set the reference when doing the triggering the sample.

```C
typedef struct {
    gpio_t pin;                 
    adc_ref_t ref;               
    uint8_t ch;     
} adc_conf_t;

...

static const adc_conf_t adc_conf[] = {
    { GPIO_PIN(ADC_PORT, 0), ADC_REF_AVCC, 0 },
    { GPIO_PIN(ADC_PORT, 1), ADC_REF_AVCC, 1 },
    { GPIO_PIN(ADC_PORT, 2), ADC_REF_AVCC, 2 },
    ...
}
```

This also simplifies the `adc_init` and `adc_sample` code.

Reply:
In general a good idea, specifically to have the reference voltage configurable. What I don't like is the duplication in `pin` and `ch` because they have a direct dependency and ADC_PORT is fixed for all anyway.

Reply:
but maybe we should get general ADC support first and fix this later on. A first step could also be to make the ref voltage configurable via define macro (though it applies to all pins then).